rero_ils/modules/acq_invoices/api.py

```diff
@@ -0,0 +1,147 @@
+               dbcommit=False, reindex=False, **kwargs):
+        """Create acquisition invoice record."""
+        cls._acquisition_invoice_build_org_ref(data)
+        cls._build_total_amount_of_invoice(data)
+        record = super(AcquisitionInvoice, cls).create(
+            data, id_, delete_pid, dbcommit, reindex, **kwargs)
+        return record
+
+    def update(self, data, dbcommit=True, reindex=True):
+        """Update Acquisition Invoice record."""
+        self._build_total_amount_of_invoice(data)
+        super(AcquisitionInvoice, self).update(data, dbcommit, reindex)
+        return self
+
+    @classmethod
+    def _build_total_amount_of_invoice(cls, data):
+        """Build total amount for invoice."""
+        invoice_price = 0
+        for idx, item in enumerate(data.get('invoice_items')):
+            # build total price for each invoice line item
+            invoiceLine = InvoiceLine(item)
+            data['invoice_items'][idx]['total_price'] = invoiceLine.total_price
+            invoice_price += data['invoice_items'][idx]['total_price']
+
+        # check if discount percentage
+        if data.get('invoice_discount', {}).get('percentage'):
+            invoice_price -= cls._calculate_percentage_discount(
+                invoice_price, data.get('invoice_discount').get('percentage'))
+        # check if discount amount
+        if data.get('invoice_discount', {}).get('amount'):
+            invoice_price -= data.get('invoice_discount').get('amount')
+        # set invoice price
+        data['invoice_price'] = invoice_price
+
+    @classmethod
+    def _calculate_percentage_discount(cls, amount, percentage):
+        """Calculate discount percentage of invoice."""
+        return amount * percentage / 100
+
+    @classmethod
+    def _acquisition_invoice_build_org_ref(cls, data):
+        """Build $ref for the organisation of the acquisition invoice."""
+        library_pid = data.get('library', {}).get('pid')
+        if not library_pid:
+            library_pid = data.get('library').get(
+                '$ref').split('libraries/')[1]
+        org_pid = Library.get_record_by_pid(library_pid).organisation_pid
+        base_url = current_app.config.get('RERO_ILS_APP_BASE_URL')
+        url_api = '{base_url}/api/{doc_type}/{pid}'
+        org_ref = {
+            '$ref': url_api.format(
+                base_url=base_url,
+                doc_type='organisations',
+                pid=org_pid or cls.organisation_pid)
+        }
+        data['organisation'] = org_ref
+
+    @property
+    def organisation_pid(self):
+        """Shortcut for acquisition invoice organisation pid."""
+        return self.replace_refs().get('organisation').get('pid')
+
+    @property
+    def library_pid(self):
+        """Shortcut for acquisition order library pid."""
+        return self.replace_refs().get('library').get('pid')
+
+    @property
+    def vendor_pid(self):
+        """Shortcut for acquisition order library pid."""
```

Comment:
"""Shortcut for acquisition order vendor pid."""

Reply:
done